File: Source/JavaScriptCore/builtins/SetPrototype.js
/*
 * Copyright (C) 2016 Yusuke Suzuki <utatane.tea@gmail.com>.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

function forEach(callback /*, thisArg */)
{
    "use strict";

    if (!@isSet(this))
        @throwTypeError("Set operation called on non-Set object");

    if (!@isCallable(callback))
        @throwTypeError("Set.prototype.forEach callback must be a function");

    var thisArg = @argument(1);
    var bucket = @setBucketHead(this);

    do {
        bucket = @setBucketNext(bucket);
        if (bucket === @sentinelSetBucket)
            break;
        var key = @setBucketKey(bucket);
        callback.@call(thisArg, key, key, this);
    } while (true);
}

// https://tc39.es/proposal-set-methods/#sec-getsetrecord (steps 1-7)
@linkTimeConstant
@alwaysInline
function getSetSizeAsInt(other)
{
    if (!@isObject(other))
        @throwTypeError("Set operation expects first argument to be an object");

    var size = @toNumber(other.size);
    if (size !== size) // is NaN?
        @throwTypeError("Set operation expects first argument to have non-NaN 'size' property");

    var sizeInt = @toIntegerOrInfinity(size);
    if (sizeInt < 0)
        @throwRangeError("Set operation expects first argument to have non-negative 'size' property");

    return sizeInt;
}

function union(other)
{
    "use strict";

    if (!@isSet(this))
        @throwTypeError("Set operation called on non-Set object");

    // Get Set Record
    var size = @getSetSizeAsInt(other); // unused but @getSetSizeAsInt call is observable

    var has = other.has;
    if (!@isCallable(has))
        @throwTypeError("Set.prototype.union expects other.has to be callable");

    var keys = other.keys;
    if (!@isCallable(keys))
        @throwTypeError("Set.prototype.union expects other.keys to be callable");

    var iterator = keys.@call(other, keys);
    var wrapper = {
        @@iterator: function () { return iterator; }
    };

    var result = @setClone(this);
    for (var key of wrapper)
        result.@add(key);

    return result;
}

function intersection(other)
{
    "use strict";

    if (!@isSet(this))
        @throwTypeError("Set operation called on non-Set object");

    // Get Set Record
    var size = @getSetSizeAsInt(other);

    var has = other.has;
    if (!@isCallable(has))
        @throwTypeError("Set.prototype.intersection expects other.has to be callable");

    var keys = other.keys;
    if (!@isCallable(keys))
        @throwTypeError("Set.prototype.intersection expects other.keys to be callable");

    var result = new @Set();
    if (this.@size <= size) {
        var bucket = @setBucketHead(this);

        do {
            bucket = @setBucketNext(bucket);
            if (bucket === @sentinelSetBucket)
                break;
            var key = @setBucketKey(bucket);
            if (has.@call(other, key))
                result.@add(key);
        } while (true);
    } else {
        var iterator = keys.@call(other, keys);
        var wrapper = {
            @@iterator: function () { return iterator; }
        };

        for (var key of wrapper) {
            if (this.@has(key))
                result.@add(key);
        }
    }

    return result;
}

function difference(other)
{
    "use strict";

    if (!@isSet(this))
        @throwTypeError("Set operation called on non-Set object");

    // Get Set Record
    var size = @getSetSizeAsInt(other);

    var has = other.has;
    if (!@isCallable(has))
        @throwTypeError("Set.prototype.difference expects other.has to be callable");

    var keys = other.keys;
    if (!@isCallable(keys))
        @throwTypeError("Set.prototype.difference expects other.keys to be callable");

    var result = @setClone(this);
    if (this.@size <= size) {
        var bucket = @setBucketHead(this);

        while (true) {
            bucket = @setBucketNext(bucket);
            if (bucket === @sentinelSetBucket)
                break;
            var key = @setBucketKey(bucket);
            if (has.@call(other, key))
                result.@delete(key);
        }
    } else {
        var iterator = keys.@call(other, keys);
        var wrapper = {
            @@iterator: function () { return iterator; }
        };

        for (var key of wrapper) {
            if (this.@has(key))
                result.@delete(key);
        }
    }

    return result;
}

function symmetricDifference(other)
{
    "use strict";

    if (!@isSet(this))
        @throwTypeError("Set operation called on non-Set object");

    // Get Set Record
    var size = @getSetSizeAsInt(other); // unused but @getSetSizeAsInt call is observable

    var has = other.has;
    if (!@isCallable(has))
        @throwTypeError("Set.prototype.symmetricDifference expects other.has to be callable");

    var keys = other.keys;
    if (!@isCallable(keys))
        @throwTypeError("Set.prototype.symmetricDifference expects other.keys to be callable");

    var iterator = keys.@call(other, keys);
    var wrapper = {
        @@iterator: function () { return iterator; }
    };

    var result = @setClone(this);
    for (var key of wrapper) {
        if (result.@has(key))
            result.@delete(key);
        else
            result.@add(key);
    }

    return result;
}

function isSubsetOf(other)
{
    "use strict";

    if (!@isSet(this))
        @throwTypeError("Set operation called on non-Set object");

    // Get Set Record
    var size = @getSetSizeAsInt(other);

    var has = other.has;
    if (!@isCallable(has))
        @throwTypeError("Set.prototype.isSubsetOf expects other.has to be callable");

    var keys = other.keys;
    if (!@isCallable(keys))
        @throwTypeError("Set.prototype.isSubsetOf expects other.keys to be callable");

    if (this.@size > size)
        return false;

    var bucket = @setBucketHead(this);

    do {
        bucket = @setBucketNext(bucket);
        if (bucket === @sentinelSetBucket)
            break;
        var key = @setBucketKey(bucket);
        if (!has.@call(other, key))
            return false;
    } while (true);

    return true;
}

function isSupersetOf(other)
{
    "use strict";

    if (!@isSet(this))
        @throwTypeError("Set operation called on non-Set object");

    // Get Set Record
    var size = @getSetSizeAsInt(other);

    var has = other.has;
    if (!@isCallable(has))
        @throwTypeError("Set.prototype.isSupersetOf expects other.has to be callable");

    var keys = other.keys;
    if (!@isCallable(keys))
        @throwTypeError("Set.prototype.isSupersetOf expects other.keys to be callable");

    if (this.@size < size)
        return false;

    var iterator = keys.@call(other, keys);
    var wrapper = {
        @@iterator: function () { return iterator; }
    };

    for (var key of wrapper) {
        if (!this.@has(key))
            return false;
    }
    return true;
}

function isDisjointFrom(other)
{
    "use strict";

    if (!@isSet(this))
        @throwTypeError("Set operation called on non-Set object");

    // Get Set Record
    var size = @getSetSizeAsInt(other);

    var has = other.has;
    if (!@isCallable(has))
        @throwTypeError("Set.prototype.isDisjointFrom expects other.has to be callable");

    var keys = other.keys;
    if (!@isCallable(keys))
        @throwTypeError("Set.prototype.isDisjointFrom expects other.keys to be callable");

    if (this.@size <= size) {
        var bucket = @setBucketHead(this);

        do {
            bucket = @setBucketNext(bucket);
            if (bucket === @sentinelSetBucket)
                break;
            var key = @setBucketKey(bucket);
            if (has.@call(other, key))
                return false;
        } while (true);
    } else {
        var iterator = keys.@call(other, keys);
        var wrapper = {
            @@iterator: function () { return iterator; }
        };

        for (var key of wrapper) {
            if (this.@has(key))
                return false;
        }
    }

    return true;
}

File: Safari-611.1.10/Source/JavaScriptCore/runtime/SetPrototype.cpp
/*
 * Copyright (C) 2013-2019 Apple Inc. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"
#include "SetPrototype.h"

#include "BuiltinNames.h"
#include "JSCInlines.h"
#include "JSSet.h"
#include "JSSetIterator.h"

#include "SetPrototype.lut.h"

namespace JSC {

const ClassInfo SetPrototype::s_info = { "Set", &Base::s_info, &setPrototypeTable, nullptr, CREATE_METHOD_TABLE(SetPrototype) };

/* Source for SetIteratorPrototype.lut.h
@begin setPrototypeTable
  forEach   JSBuiltin  DontEnum|Function 0
@end
*/

static JSC_DECLARE_HOST_FUNCTION(setProtoFuncAdd);
static JSC_DECLARE_HOST_FUNCTION(setProtoFuncClear);
static JSC_DECLARE_HOST_FUNCTION(setProtoFuncDelete);
static JSC_DECLARE_HOST_FUNCTION(setProtoFuncHas);
static JSC_DECLARE_HOST_FUNCTION(setProtoFuncValues);
static JSC_DECLARE_HOST_FUNCTION(setProtoFuncEntries);

static JSC_DECLARE_HOST_FUNCTION(setProtoFuncSize);

void SetPrototype::finishCreation(VM& vm, JSGlobalObject* globalObject)
{
    Base::finishCreation(vm);
    ASSERT(inherits(vm, info()));

    JSC_NATIVE_INTRINSIC_FUNCTION_WITHOUT_TRANSITION(vm.propertyNames->add, setProtoFuncAdd, static_cast<unsigned>(PropertyAttribute::DontEnum), 1, JSSetAddIntrinsic);
    JSC_NATIVE_FUNCTION_WITHOUT_TRANSITION(vm.propertyNames->clear, setProtoFuncClear, static_cast<unsigned>(PropertyAttribute::DontEnum), 0);
    JSC_NATIVE_FUNCTION_WITHOUT_TRANSITION(vm.propertyNames->deleteKeyword, setProtoFuncDelete, static_cast<unsigned>(PropertyAttribute::DontEnum), 1);
    JSC_NATIVE_INTRINSIC_FUNCTION_WITHOUT_TRANSITION(vm.propertyNames->has, setProtoFuncHas, static_cast<unsigned>(PropertyAttribute::DontEnum), 1, JSSetHasIntrinsic);
    JSC_NATIVE_INTRINSIC_FUNCTION_WITHOUT_TRANSITION(vm.propertyNames->builtinNames().entriesPublicName(), setProtoFuncEntries, static_cast<unsigned>(PropertyAttribute::DontEnum), 0, JSSetEntriesIntrinsic);
    JSC_NATIVE_INTRINSIC_FUNCTION_WITHOUT_TRANSITION(vm.propertyNames->builtinNames().hasPrivateName(), setProtoFuncHas, static_cast<unsigned>(PropertyAttribute::DontEnum), 1, JSSetHasIntrinsic);
    JSC_NATIVE_INTRINSIC_FUNCTION_WITHOUT_TRANSITION(vm.propertyNames->builtinNames().addPrivateName(), setProtoFuncAdd, static_cast<unsigned>(PropertyAttribute::DontEnum), 1, JSSetAddIntrinsic);

    JSFunction* values = JSFunction::create(vm, globalObject, 0, vm.propertyNames->builtinNames().valuesPublicName().string(), setProtoFuncValues, JSSetValuesIntrinsic);
    putDirectWithoutTransition(vm, vm.propertyNames->builtinNames().valuesPublicName(), values, static_cast<unsigned>(PropertyAttribute::DontEnum));
    putDirectWithoutTransition(vm, vm.propertyNames->builtinNames().keysPublicName(), values, static_cast<unsigned>(PropertyAttribute::DontEnum));
    putDirectWithoutTransition(vm, vm.propertyNames->iteratorSymbol, values, static_cast<unsigned>(PropertyAttribute::DontEnum));
    JSC_TO_STRING_TAG_WITHOUT_TRANSITION();

    JSC_NATIVE_GETTER_WITHOUT_TRANSITION(vm.propertyNames->size, setProtoFuncSize, PropertyAttribute::DontEnum | PropertyAttribute::Accessor);

    globalObject->installSetPrototypeWatchpoint(this);
}

ALWAYS_INLINE static JSSet* getSet(JSGlobalObject* globalObject, JSValue thisValue)
{
    VM& vm = globalObject->vm();
    auto scope = DECLARE_THROW_SCOPE(vm);

    if (UNLIKELY(!thisValue.isCell())) {
        throwVMError(globalObject, scope, createNotAnObjectError(globalObject, thisValue));
        return nullptr;
    }
    auto* set = jsDynamicCast<JSSet*>(vm, thisValue.asCell());
    if (LIKELY(set))
        return set;
    throwTypeError(globalObject, scope, "Set operation called on non-Set object"_s);
    return nullptr;
}

JSC_DEFINE_HOST_FUNCTION(setProtoFuncAdd, (JSGlobalObject* globalObject, CallFrame* callFrame))
{
    JSValue thisValue = callFrame->thisValue();
    JSSet* set = getSet(globalObject, thisValue);
    if (!set)
        return JSValue::encode(jsUndefined());
    set->add(globalObject, callFrame->argument(0));
    return JSValue::encode(thisValue);
}

JSC_DEFINE_HOST_FUNCTION(setProtoFuncClear, (JSGlobalObject* globalObject, CallFrame* callFrame))
{
    JSSet* set = getSet(globalObject, callFrame->thisValue());
    if (!set)
        return JSValue::encode(jsUndefined());
    set->clear(globalObject);
    return JSValue::encode(jsUndefined());
}

JSC_DEFINE_HOST_FUNCTION(setProtoFuncDelete, (JSGlobalObject* globalObject, CallFrame* callFrame))
{
    JSSet* set = getSet(globalObject, callFrame->thisValue());
    if (!set)
        return JSValue::encode(jsUndefined());
    return JSValue::encode(jsBoolean(set->remove(globalObject, callFrame->argument(0))));
}

JSC_DEFINE_HOST_FUNCTION(setProtoFuncHas, (JSGlobalObject* globalObject, CallFrame* callFrame))
{
    JSSet* set = getSet(globalObject, callFrame->thisValue());
    if (!set)
        return JSValue::encode(jsUndefined());
    return JSValue::encode(jsBoolean(set->has(globalObject, callFrame->argument(0))));
}

JSC_DEFINE_HOST_FUNCTION(setProtoFuncSize, (JSGlobalObject* globalObject, CallFrame* callFrame))
{
    JSSet* set = getSet(globalObject, callFrame->thisValue());
    if (!set)
        return JSValue::encode(jsUndefined());
    return JSValue::encode(jsNumber(set->size()));
}

inline JSValue createSetIteratorObject(JSGlobalObject* globalObject, CallFrame* callFrame, IterationKind kind)
{
    VM& vm = globalObject->vm();
    JSValue thisValue = callFrame->thisValue();
    JSSet* set = getSet(globalObject, thisValue);
    if (!set)
        return jsUndefined();
    return JSSetIterator::create(vm, globalObject->setIteratorStructure(), set, kind);
}

JSC_DEFINE_HOST_FUNCTION(setProtoFuncValues, (JSGlobalObject* globalObject, CallFrame* callFrame))
{
    return JSValue::encode(createSetIteratorObject(globalObject, callFrame, IterationKind::Values));
}

JSC_DEFINE_HOST_FUNCTION(setProtoFuncEntries, (JSGlobalObject* globalObject, CallFrame* callFrame))
{
    return JSValue::encode(createSetIteratorObject(globalObject, callFrame, IterationKind::Entries));
}

}
